src/core/algorithms/nar/des/encoded_value_range.h

```diff
@@ -0,0 +1,112 @@
+#include "rng.h"
+
+namespace algos::des {
+
+class EncodedValueRange {
+public:
+    static size_t const kFieldCount = 4;
+    double permutation;
+    double threshold;
+    double bound1;
+    double bound2;
+
+    // TODO: remove code duplication here
+    double& operator[](size_t index) {
+        switch (index) {
+            case 0:
+                return permutation;
+                break;
+            case 1:
+                return threshold;
+                break;
+            case 2:
+                return bound1;
+                break;
+            case 3:
+                return bound2;
+                break;
+            default:
+                throw std::out_of_range("Index out of range for value range.");
+        }
+    }
+
+    double const& operator[](size_t index) const {
+        switch (index) {
+            case 0:
+                return permutation;
+                break;
+            case 1:
+                return threshold;
+                break;
+            case 2:
+                return bound1;
+                break;
+            case 3:
+                return bound2;
+                break;
+            default:
+                throw std::out_of_range("Index out of range for value range.");
+        }
+    }
+
+    std::shared_ptr<model::ValueRange> Decode(std::shared_ptr<model::ValueRange> domain) const {
+        switch (domain->GetTypeId()) {
+            case model::TypeId::kInt: {
+                std::shared_ptr<model::IntValueRange> int_domain =
+                        std::static_pointer_cast<model::IntValueRange>(domain);
+                model::Int span = int_domain->upper_bound - int_domain->lower_bound;
+                model::Int resulting_lower = int_domain->lower_bound + span * this->bound1;
+                model::Int resulting_upper = int_domain->lower_bound + span * this->bound2;
+                if (resulting_lower > resulting_upper) {
+                    std::swap(resulting_lower, resulting_upper);
+                }
+                return std::make_shared<model::IntValueRange>(
+                        model::IntValueRange(resulting_lower, resulting_upper));
+            }
+            case model::TypeId::kDouble: {
+                std::shared_ptr<model::DoubleValueRange> double_domain =
+                        std::static_pointer_cast<model::DoubleValueRange>(domain);
+                model::Double span = double_domain->upper_bound - double_domain->lower_bound;
+                model::Double resulting_lower = double_domain->lower_bound + span * this->bound1;
+                model::Double resulting_upper = double_domain->lower_bound + span * this->bound2;
+                if (resulting_lower > resulting_upper) {
+                    std::swap(resulting_lower, resulting_upper);
+                }
+                return std::make_shared<model::DoubleValueRange>(
+                        model::DoubleValueRange(resulting_lower, resulting_upper));
+            }
```

Comment:
Maybe it's better to use private template method here?
```
template <typename T, typename RangeT>
    std::shared_ptr<RangeT> DecodeTypedValueRange(std::shared_ptr<model::ValueRange> const& domain) const {
        std::shared_ptr<RangeT> int_domain = std::static_pointer_cast<RangeT>(domain);
        T span = int_domain->upper_bound - int_domain->lower_bound;
        T resulting_lower = int_domain->lower_bound + span * this->bound1;
        T resulting_upper = int_domain->lower_bound + span * this->bound2;
        if (resulting_lower > resulting_upper) {
            std::swap(resulting_lower, resulting_upper);
        }
        return std::make_shared<RangeT>(resulting_lower, resulting_upper);
    }
```
and call it here:
```suggestion
            case TypeId::kInt: {
            return DecodeTypedValueRange<Int, IntValueRange>(domain);
        }
        case TypeId::kDouble: {
            return DecodeTypedValueRange<Double, DoubleValueRange>(domain);
        }
```
Also you can define `Mixed` case as template specialization:
```
template <>
std::shared_ptr<model::StringValueRange>
EncodedValueRange::DecodeTypedValueRange<model::String, model::StringValueRange>
(std::shared_ptr<model::ValueRange> const& domain) const {
    using namespace model;
    std::shared_ptr<StringValueRange> string_domain =
            std::static_pointer_cast<StringValueRange>(domain);
    std::vector<String> string_vector = string_domain->domain;
    size_t span = string_vector.size();
    // upper_bound is not used, resulting NARs bind categorical values with a single
    // value.
    String result;
    if (bound1 == 1.0) {
        result = string_vector.back();
    } else {
        result = string_vector[(size_t)(span * this->bound1)];
    }
    return std::make_shared<StringValueRange>(StringValueRange(result));
}
```
Note that specialization should appear **after** template itself and **before** first instantiation. The easier way to do this is to place template specialization and `EncodeValueRange::Decode` definition in `.cpp` file -- in that order

Reply:
also we should avoid using c-style casts: "`(size_t)`"